bin/packages/check-build-type-declaration-files.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * The purpose of this script is to double-check certain published build types.
+ * When we translate JSDoc into types that we wish to publish, we occasionally
+ * disable the type checker for JS files. Unfortunately, this can result in invalid
+ * index.d.ts files being published.
+ *
+ * This script verifies the publishd index.d.ts file for every package which both
+ * builds types and also sets checkJs to false in its tsconfig.json. This is done
+ * by running `tsc --noEmit` on the $package/build-types/index.d.ts file. This also
+ * verifies everything index.d.ts references, which essentially checks the entire
+ * public api of the type declarations for that package.
+ *
+ * @see https://github.com/WordPress/gutenberg/pull/49650 for more discussion.
+ */
+
+/**
+ * External dependencies
+ */
+const fs = require( 'fs' ).promises;
+const path = require( 'path' );
+const { exec } = require( 'child_process' );
+const chalk = require( 'chalk' );
+
+// Tsconfig can be JSON5, which accepts comments, which need to be stripped so we can parse the JSON.
+function stripJsonComments( jsonString ) {
+	const lines = jsonString.split( '\n' );
+	const cleanedLines = lines.map( ( line ) => line.replace( /\/\/.*$/, '' ) );
+	return cleanedLines.join( '\n' );
+}
+
+async function readTsConfig( tsconfigPath ) {
+	const tsconfigRaw = await fs.readFile( tsconfigPath, 'utf-8' );
+	return JSON.parse( stripJsonComments( tsconfigRaw ) );
+}
+
+async function typecheckDeclarations( file ) {
+	return new Promise( ( resolve, reject ) => {
+		exec( `npx tsc --noEmit ${ file }`, ( error, stdout, stderr ) => {
+			if ( error ) {
+				reject( { file, error, stderr, stdout } );
+			} else {
+				resolve( { file, stdout } );
+			}
+		} );
+	} );
+}
+
+// Returns the path to the build-types declaration file for a package if it exists.
+// Throws an error and exits the script otherwise.
+async function getDecFile( packagePath ) {
+	const decFile = path.join( packagePath, 'build-types', 'index.d.ts' );
+	try {
+		await fs.access( decFile );
+		return decFile;
+	} catch ( err ) {
+		if ( err.code !== 'ENOENT' ) {
+			throw err;
+		}
+		console.error(
+			`This declaration file should exist. You may need to run tsc again: ${ decFile }`
+		);
+		process.exit( 1 );
+	}
+}
```

Comment:
in this function I think `fileExistsSync` is probably still appropriate since our purpose is to check the presence of the file but not read it in.

there is another approach we could take ~if we can use `fs/promises`~ (oops look like you already made that change), which is that this function could return a `FileHandle`, but still fail with `ENOENT` when the file doesn't exist.

passing that `FileHandle` along means we can later call `fsPromises.readFile` with the handle, which removes the need to pass along the filename again.

none of this really matters, but I'm passing along the comment in case you find it useful.

```ts
async function getDecFile( packagePage ) {
	const decFile = …;

	try {
		return fs.open( decFile, 'r' );
	} catch ( e ) {
		if ( e.code !== 'ENOENT' ) {
			throw e;
		}

		console.error( … );
		process.exit( 1 );
	}
}
```

At the point we have a `FileHandle` I don't think it's possible to get `ENOENT` again because the system will keep the file open in memory even if it's deleted by another process on disk.

given that `readTsConfig` is only used in determining if `checkJs` is `true`, we could inline that as well as the error handling into `readTsConfig` and then the main `const declarations = ( … )` would have no more error-handling

```ts
async function packageChecksJs( packagePath ) {
	const configPath = path.join( packagePath, 'tsconfig.json' );

	try {
		const tsconfigRaw = await fs.readFile( configPath, 'utf-8' );
	} catch ( e ) {
		if ( e.code !== 'ENOENT' ) {
			throw e;
		}

		// No tsconfig means no checkJs
		return false;
	}

	const config = JSON.parse( stripJsonComments( tsconfigRaw ) );
	return config.compilerOptions?.checkJs === true;
}
```

```ts
const declarations = ( await Promise.all(
	packageDirs.map( async ( pkg ) =>
		( await packageChecksJs( pkg ) )
			? null
			: getDecFile( pkg )
	)
) ).filter( Boolean );
```